In [1]:
from keras.models import Sequential
from keras.layers import Dense
from keras import regularizers

import numpy as np
# fix random seed for reproducibility
np.random.seed(42)

import pandas as pd

Using TensorFlow backend.


In [2]:
#Folder for the dataset
datasetFolder = '/home/carnd/dbpedia2016/all4_1x15/dataset/'

#Number of files
numberOfFiles = 2555

#Test split
testSplit=0.25

In [3]:
def load_data(datasetFolder, datasetXFile, datasetYFile, wrap=True):
    #print('Loading {} & {}'.format(datasetXFile, datasetYFile))
    # load file
    with open(datasetFolder + datasetXFile, "r") as f:
        head = f.readline()
        cols = head.split(',')
        numberOfCols = len(cols)
        #print(numberOfCols)
        numberOfRows=0
        for line in f:
            numberOfRows+=1
        f.close()
    
    #print('{} x {}'.format(numberOfRows,numberOfCols))
    if(wrap==True):
        maxY = 1000
    else:
        maxY = numberOfCols-1
    dataX = np.zeros([numberOfRows,maxY],np.int8)
    with open(datasetFolder + datasetXFile, "r") as f:
        head = f.readline()
        rowCounter=0
        for line in f:
            row=line.split(', ')
            for i in range(1, len(row)):
                if(int(row[i])<=0):
                    continue;
                dataX[rowCounter][(int(row[i])-1)%maxY]=1 + ((int(row[i])-1)//maxY)
                #if((1 + ((int(row[i])-1)//maxY))>1):
                #    print("{} data[{}][{}] = {}".format(int(row[i])-1, rowCounter,(int(row[i])-1)%maxY,1 + ((int(row[i])-1)//maxY)))
            rowCounter+=1
        f.close()
        
    dataY=pd.read_csv(datasetFolder + datasetYFile)

    # delete the id fields
    #del dataX['id']
    del dataY['id']

    return dataX, dataY.as_matrix()

In [4]:
dataX, dataY = load_data(datasetFolder,'datasetX_1231.csv', 'datasetY_1231.csv',wrap=True)

In [5]:
print(dataX.shape)
print(dataX[0:5])

(1241, 1000)
[[0 1 0 ..., 0 0 0]
 [0 0 0 ..., 0 0 0]
 [0 0 0 ..., 0 0 0]
 [0 0 0 ..., 0 0 0]
 [0 1 0 ..., 0 0 0]]


In [6]:
print(dataY.shape)
print(dataY[0:5])

(1241, 539)
[[0 0 0 ..., 0 0 0]
 [0 0 0 ..., 0 0 0]
 [0 0 0 ..., 0 0 0]
 [0 0 0 ..., 0 0 0]
 [0 0 0 ..., 0 0 0]]


In [7]:
print("Input Features: {}".format(dataX.shape[1]))
print("Output Classes: {}".format(dataY.shape[1]))

Input Features: 1000
Output Classes: 539


In [8]:
# create model
# logisticRegression = Sequential(name='Simple Logistic Regression')
# logisticRegression.add(Dense(dataY.shape[1], input_dim=dataX.shape[1], activation='sigmoid', init='glorot_uniform'))

# simpleModel = Sequential(name='2 Fully Connected Layers')
# simpleModel.add(Dense(1024, input_dim=dataX.shape[1], activation='relu', init='glorot_uniform'))
# simpleModel.add(Dense(dataY.shape[1], activation='sigmoid', init='glorot_uniform'))

deepModel = Sequential(name='Deep Model (5 Dense Layers)')
deepModel.add(Dense(2048, input_dim=dataX.shape[1], activation='tanh', init='glorot_uniform'))
deepModel.add(Dense(2048, activation='elu', init='glorot_uniform'))
deepModel.add(Dense(1024, activation='tanh', init='glorot_uniform'))
deepModel.add(Dense(1024, activation='elu', init='glorot_uniform'))
deepModel.add(Dense(768, activation='elu', init='glorot_uniform'))
deepModel.add(Dense(dataY.shape[1], activation='sigmoid', init='glorot_uniform'))

models = [deepModel]

In [9]:
# Compile model
import keras.backend as K

def count_predictions(y_true, y_pred):
    true_positives = K.sum(K.round(K.clip(y_true * y_pred, 0, 1)))
    predicted_positives = K.sum(K.round(K.clip(y_pred, 0, 1)))
    possible_positives = K.sum(K.round(K.clip(y_true, 0, 1)))
    return true_positives, predicted_positives, possible_positives

def f1score(y_true, y_pred):
    true_positives, predicted_positives, possible_positives = count_predictions(y_true, y_pred)
    precision = true_positives / (predicted_positives + K.epsilon())
    recall = true_positives / (possible_positives + K.epsilon())
    f1score = 2.0 * precision * recall / (precision+recall+ K.epsilon())
    return f1score

def fBetaScore(y_true, y_pred, beta):
    true_positives, predicted_positives, possible_positives = count_predictions(y_true, y_pred)
    precision = true_positives / (predicted_positives + K.epsilon())
    recall = true_positives / (possible_positives + K.epsilon())
    f1score = (1+(beta*beta)) * precision * recall / ((beta*beta*precision)+recall+ K.epsilon())
    return f1score

for model in models:
    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=[f1score])

In [10]:
def fit_data(model, dataX, dataY):
    # Fit the model
    #model.fit(dataX, dataY, nb_epoch=5, verbose=2, batch_size=256)
    return model.train_on_batch(dataX, dataY)

In [11]:
def countPredictions(y_true, y_pred):
    true_positives = np.sum(np.round(y_pred*y_true))
    predicted_positives = np.sum(np.round(y_pred))
    possible_positives = np.sum(y_true)
    return true_positives, predicted_positives, possible_positives

In [12]:
#Randomize the list of numbers so we can split train and test dataset
listOfFiles=list(range(1,numberOfFiles+1))
import random
random.shuffle(listOfFiles)
splitIndex=int((1-testSplit)*numberOfFiles)

numberOfEons = 15
for eon in range(0, numberOfEons):
    print('{}. Eon {}/{}'.format(eon+1,eon+1, numberOfEons))
    for trainIndex in range(0,splitIndex):
        dataX, dataY = load_data(datasetFolder,'datasetX_{}.csv'.format(listOfFiles[trainIndex]), 'datasetY_{}.csv'.format(listOfFiles[trainIndex]))
        for model in models:
            #print('Model = {}'.format(model.name))
            #model.fit(dataX, dataY, nb_epoch=5, verbose=0, batch_size=512)
            loss, f1score=fit_data(model,dataX, dataY)
            print('Learning for file {} / {} : datasetX/Y_{}\t\tloss={:.4f} f1score={:.4f}'.format(trainIndex+1, splitIndex, listOfFiles[trainIndex], loss, f1score), end='\r')
        
#             if((trainIndex+1)%10==0):
#                 sc=model.test_on_batch(dataX,dataY)
#                 print('Learning for file {} / {} : datasetX/Y_{}...'.format(trainIndex+1, splitIndex, listOfFiles[trainIndex]))
#                 print('   Model = {} \t loss = {:.4f} \t f1-score = {:.4f}'.format(model.name, sc[0], sc[1]))
#                 print('==========================================')
             
        
    counts = {} 
    for model in models:
        counts[model.name] = {'true_positives':0, 'predicted_positives':0, 'possible_positives':0}
    
    for testIndex in range(splitIndex, numberOfFiles):
        dataX, dataY = load_data(datasetFolder,'datasetX_{}.csv'.format(listOfFiles[testIndex]), 'datasetY_{}.csv'.format(listOfFiles[testIndex]))
        for model in models:
            predY=model.predict_on_batch(dataX)
            true_positives, predicted_positives, possible_positives = countPredictions(dataY, predY)
            counts[model.name]['true_positives'] += true_positives
            counts[model.name]['predicted_positives'] += predicted_positives
            counts[model.name]['possible_positives'] += possible_positives
            print ('Testing for file {} / {} : datasetX/Y_{} - true +ve:{}  pred +ve:{} possible +ve:{}'.format(testIndex+1, numberOfFiles, listOfFiles[testIndex], true_positives,predicted_positives,possible_positives), end='\r')
    
    for model in models:
        count = counts[model.name]
        precision = (count['true_positives'])/(count['predicted_positives']+0.0001)
        recall = (count['true_positives'])/(count['possible_positives']+0.0001)
        f1score = 2.0 * precision * recall / (precision+recall+0.0001)
        print('- Model = {} \t f1-score = {:.4f}\t precision = {:.4f} \t recall = {:.4f'.format(model.name, f1score, precision, recall))

1. Eon 1/15
- Model = Deep Model (5 Dense Layers) 	 f1-score = 0.8047	 precision = 0.8835 	 recall = 0.7388                          

2. Eon 2/15
- Model = Deep Model (5 Dense Layers) 	 f1-score = 0.8374	 precision = 0.9029 	 recall = 0.7808                          

3. Eon 3/15
- Model = Deep Model (5 Dense Layers) 	 f1-score = 0.8516	 precision = 0.9000 	 recall = 0.8083                          

4. Eon 4/15
- Model = Deep Model (5 Dense Layers) 	 f1-score = 0.8593	 precision = 0.9067 	 recall = 0.8167                          

5. Eon 5/15
- Model = Deep Model (5 Dense Layers) 	 f1-score = 0.8624	 precision = 0.9134 	 recall = 0.8169                          

6. Eon 6/15
- Model = Deep Model (5 Dense Layers) 	 f1-score = 0.8656	 precision = 0.9181 	 recall = 0.8189                          

7. Eon 7/15
- Model = Deep Model (5 Dense Layers) 	 f1-score = 0.8690	 precision = 0.9168 	 recall = 0.8260                          

8. Eon 8/15
- Model = Deep Model (5 Dense Layers) 	 f1-

==================================================
# all4_1x15
==================================================
1. Eon 1/15
 - Model = Deep Model (5 Dense Layers) 	 f1-score = 0.8047	 precision = 0.8835 	 recall = 0.7388  
2. Eon 2/15:
 - Model = Deep Model (5 Dense Layers) 	 f1-score = 0.8374	 precision = 0.9029 	 recall = 0.7808                   
3. Eon 3/15:
 - Model = Deep Model (5 Dense Layers) 	 f1-score = 0.8516	 precision = 0.9000 	 recall = 0.8083                   
4. Eon 4/15:
 - Model = Deep Model (5 Dense Layers) 	 f1-score = 0.8593	 precision = 0.9067 	 recall = 0.8167                   
5. Eon 5/15:
 - Model = Deep Model (5 Dense Layers) 	 f1-score = 0.8624	 precision = 0.9134 	 recall = 0.8169                   
6. Eon 6/15:
 - Model = Deep Model (5 Dense Layers) 	 f1-score = 0.8656	 precision = 0.9181 	 recall = 0.8189                   
7. Eon 7/15:
 - Model = Deep Model (5 Dense Layers) 	 f1-score = 0.8690	 precision = 0.9168 	 recall = 0.8260                   
8. Eon 8/15:
 - Model = Deep Model (5 Dense Layers) 	 f1-score = 0.8700	 precision = 0.9172 	 recall = 0.8275
9. Eon 9/15:
 - Model = Deep Model (5 Dense Layers) 	 f1-score = 0.8710	 precision = 0.9168 	 recall = 0.8296                   
10. Eon 10/15:
 - Model = Deep Model (5 Dense Layers) 	 f1-score = 0.8725	 precision = 0.9190 	 recall = 0.8305                   
11. Eon 11/15:
 - Model = Deep Model (5 Dense Layers) 	 f1-score = 0.8733	 precision = 0.9180 	 recall = 0.8329                   
12. Eon 12/15:
 - Model = Deep Model (5 Dense Layers) 	 f1-score = 0.8736	 precision = 0.9219 	 recall = 0.8302                   
13. Eon 13/15:
 - Model = Deep Model (5 Dense Layers) 	 f1-score = 0.8745	 precision = 0.9206 	 recall = 0.8328                   
14. Eon 14/15:
 - Model = Deep Model (5 Dense Layers) 	 f1-score = 0.8739	 precision = 0.9221 	 recall = 0.8305                   
15. Eon 15/15:
 - Model = Deep Model (5 Dense Layers) 	 f1-score = 0.8746	 precision = 0.9211 	 recall = 0.8327      

## 